Added SQLCipher support

BrandtM opened the pull request:
Hey. 

I added SQLCipher support (yay).  
It was really easy to implement too so this should be pretty straightforward to merge.  

Here is some important information:  
* This requires header files for SQLCipher and I think sqlite3 as well.
I had the following packages installed: sqlite3 and libsqlcipher-dev. Though I'm pretty sure only the dev package has to be installed. But it has to be installed **before** installing pysqlcipher3 via pip.  

* The passphrase for the DB has to be at least 8 characters long (it will exit with an error if it's not the case)  

* The passphrase a DB has been created with is the only passphrase that can decipher that DB  

* The key is read from a file called *"cipher_key.txt"* which resides next to the *journal.py* file. I have added it to the .gitignore file because you definitely wouldn't want your secret keys versioned. 

I'd love to hear your feedback on this.

prodicus commented on journal.py:
```diff
@@ -25,7 +25,9 @@
 path = os.getenv('HOME', os.path.expanduser('~')) + '/.tnote'
-db = SqliteDatabase(path + '/diary.db')
+
+with open('cipher_key.txt', 'r') as cipher_file:
```
On first thoughts, I think `cipher_key.txt` should not be visible to anybody. Placing it on the `~/.tnote` folder isn't any good to.

A passphrase approach combined with encryption seems viable. What's your opinion on that?

BrandtM commented:
I think we should let users enter their own cipher key for example when they start the app for the first time. This would increase security a lot more.  
Should I work on that or can you think of another approach?

prodicus commented:
That sounds good!

Waiting for your PR then :+1:

BrandtM commented:
I updated it now.  
The way it works is that you have to enter a key and a passphrase when starting the application. I implemented it in such a way that authorization is required every time it is started instead of just once.  
That way no one can snoop around and I found it a much better system anyway.  

I hope you like my implementation.

prodicus commented:
Awesome! Just one thing. Please update your branch with the master as I am having some merge conflicts. 

Will merge it then. Thanks

BrandtM commented:
Merged it.

prodicus commented:
I was testing it under python3 and it gave me this error.

```sh
tasdik at Acer in ~/Desktop/tnote on ftr-sqlcipher
$ pip freeze
args==0.1.0
clint==0.5.1
peewee==2.8.0
pycrypto==2.6.1
pysqlcipher3==1.0.1
wheel==0.24.0
(tnote)
tasdik at Acer in ~/Desktop/tnote on ftr-sqlcipher
$ ./tnote.py 
Traceback (most recent call last):
  File "/home/tasdik/.virtualenvs/tnote/lib/python3.4/site-packages/playhouse/sqlcipher_ext.py", line 59, in <module>
    from pysqlcipher import dbapi2 as sqlcipher
ImportError: No module named 'pysqlcipher'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tasdik/.virtualenvs/tnote/lib/python3.4/site-packages/playhouse/sqlcipher_ext.py", line 62, in <module>
    from pysqlcipher3 import dbapi2 as sqlcipher
  File "/home/tasdik/.virtualenvs/tnote/lib/python3.4/site-packages/pysqlcipher3/dbapi2.py", line 33, in <module>
    from pysqlcipher3._sqlite3 import *
ImportError: libsqlcipher.so.0: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./tnote.py", line 22, in <module>
    from playhouse.sqlcipher_ext import SqlCipherDatabase
  File "/home/tasdik/.virtualenvs/tnote/lib/python3.4/site-packages/playhouse/sqlcipher_ext.py", line 64, in <module>
    raise ImportError('Sqlcipher python bindings not found.')
ImportError: Sqlcipher python bindings not found.
(tnote)
tasdik at Acer in ~/Desktop/tnote on ftr-sqlcipher
$ 
```

BrandtM commented:
Did you install ```libsqlcipher-dev```?  
I am looking for more dependencies but I think this was the only one

prodicus commented:
Works now. But I am unsure how people will windows or a MAC will be able to install it. Any ideas on that?

BrandtM commented:
There are instructions on getting it done on Windows [here](https://github.com/rigglemania/pysqlcipher3) but sadly it does seem to be a bit elaborate and requires lots of tools and manual building.  
I don't know of a workaround for Windows or a package that gets everything working out of the box.

prodicus commented:
That would be undesirable for windows users. Installation should be hassle free. 

What about Mac? Is it straightforward?

We can do spin something which detects the OS and decides whether to use encryption or not.

What do you think?

BrandtM commented:
I don't know about Mac. I think it should be possible with brew but I don't know.  

I think I could implement a different encryption method for Windows. I would just encrypt/decrypt the SQL file at the start/end of the program and leave the SQL connection unencrypted.  
Would that be desirable?  

That is, at least in my opinion, a better alternative to not having encryption at all for Windows.

prodicus commented:
I am new to cryptographic terminologies, so correct me if I am wrong here @BrandtM . 

So currently we are

- encrypting the database file (using ?)
- encrypting the SQL connection (using?)

BrandtM commented:
Currently the database file is not encrypted directly. Encryption is achieved using peewee's implementation of pysqlcipher3 which results in an encrypted file.  
I will update the branch to not use encryption on Windows.  

Generally I don't like having one platform at a disadvantage in terms of features but the only alternative I see here is that users go through a fairly complicated and lengthy process of manual compilation.  

Since this is a fairly small project I think users can live without encryption on Windows.

BrandtM commented:
Done
